bgpd/bgp_route.c

```diff
@@ -4008,11 +4008,14 @@ bgp_static_update_safi (struct bgp *bgp, struct prefix *p,
 
   if ((safi == SAFI_EVPN) || (safi == SAFI_MPLS_VPN) || (safi == SAFI_ENCAP))
     {
-      if (bgp_static->igpnexthop.s_addr)
+      if (afi == AFI_IP)
         {
-          bgp_attr_extra_get (&attr)->mp_nexthop_global_in = bgp_static->igpnexthop;
+          if (!bgp_static->igpnexthop.s_addr)
+            attr.nexthop = bgp->router_id;
```

Comment:
In the same link Ivan says this:
_"When a BGP route with missing next hop is sent to BGP neighbors, the BGP next hop is set to the source IP address of the BGP session."_

I think this is what we have to do:
1 - Initialize `mp_nexthop_len` in `bgp_static_update_safi()` regardless if we have an IGP nexthop or not (the fact that we are not keeping track of IGP nexthops is a separate issue): 
```diff
--- a/bgpd/bgp_route.c
+++ b/bgpd/bgp_route.c
@@ -4086,11 +4086,8 @@ bgp_static_update_safi (struct bgp *bgp, struct prefix *p,
 
   if ((safi == SAFI_EVPN) || (safi == SAFI_MPLS_VPN) || (safi == SAFI_ENCAP))
     {
-      if (bgp_static->igpnexthop.s_addr)
-        {
-          bgp_attr_extra_get (&attr)->mp_nexthop_global_in = bgp_static->igpnexthop;
-          bgp_attr_extra_get (&attr)->mp_nexthop_len = IPV4_MAX_BYTELEN;
-        }
+      bgp_attr_extra_get (&attr)->mp_nexthop_global_in = bgp_static->igpnexthop;
+      bgp_attr_extra_get (&attr)->mp_nexthop_len = IPV4_MAX_BYTELEN;
     }
   if(afi == AFI_L2VPN)
     {
```
This way `BGP_NEXTHOP_AFI_FROM_NHLEN` will work as expected in `bgp_packet_mpattr_start()` and `nh_afi` will be set correctly.

The diff needs to be updated to handle IPv6 as well.

2 - Fix `bpacket_reformat_for_peer()` to rewrite the MP nexthop using `peer->nexthop` when `mp_nexthop_global_in` is zero.